common: assert if buffer advance length overflow

david-z opened the pull request:
Buffer advance length was defined as int, but in async msg, the real
length of data buffer was defined as unsigned, so this overflow caused
MDS crash.

Fixes: http://tracker.ceph.com/issues/36340

Signed-off-by: Zhi Zhang <zhangz.david@outlook.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

tchaikov commented:
i agree that `int` is not large enough to hold an `unsigned`. but i don't think this is the right way to fix http://tracker.ceph.com/issues/36340 , because

1. this changes the ABI, and hence breaks backward compatibility. unless we bump up the soversion.
1. i'm not sure if it's expected to have a message sizes 3.389G. could you check why the message was that large?

probably, we can have a `ceph_assert(read_len < std::numeric_limits<int>::max())` in `ProtocolV1::handle_message_data()` before passing `read_len` to `bufferlist::iterator::advance()`.

tchaikov commented:
probably https://github.com/ceph/ceph/pull/24459/commits/416bab914b80e733c1cda6ba033db9973caa09b3 is a better fix.

david-z commented:
> probably [416bab9](https://github.com/ceph/ceph/commit/416bab914b80e733c1cda6ba033db9973caa09b3) is a better fix.

Yes, this also works. 

I will close this one.

batrick commented:
> i'm not sure if it's expected to have a message sizes 3.389G. could you check why the message was that large?

@ukernel @gregsfortytwo is this really expected?

>probably 416bab9 is a better fix.

@tchaikov is this in a PR somewhere?

tchaikov commented:
@batrick #24459

david-z commented:
> i'm not sure if it's expected to have a message sizes 3.389G. could you check why the message was that large?

We didn't cache the exact OP (or message), but from the coredump, we found it came from OSD. Since this is a very old cluster upgraded from 0.94.x and it has a cephfs directory containing more than 10 million files and we haven't turned on some dir fragment optimizations, we think it might be caused by the MDS to fetch the omap.

david-z commented:
Just a suggestion :), #24459 seems changed a lot and I feel it won't be easily backported to Jewel or Luminous in a long time. So this overflow issue is still in old ceph version and there is no compatible issue to change the advance length to `int64_t` in old ceph version, could we just backport this change to old ceph version?

gregsfortytwo commented:
@batrick @tchaikov It's generally not expected to have messages that large but I think we've seen them on dirfrags that didn't have the 100k inode limit set, so they were created very large and the MDS does a single read of unlimited size, as @david-z suggests.

batrick commented:
Reopening this. I'm not sure what the right approach is but we need a fix that can be backported.

batrick commented:
@david-z can you check the coredump if the issue was caused by the MDS reading a large dirfrag?

david-z commented:
From the coredump, I can only see this message came from OSD.18 and its type was CEPH_MSG_OSD_OPREPLY. 

Pls see the details:
```
(gdb) bt full
#0  0x00007f563214423b in raise () from /lib64/libpthread.so.0
No symbol table info available.
#1  0x00007f5633269b95 in reraise_fatal (signum=6) at global/signal_handler.cc:71
        ret = <optimized out>
        buf = "\300\331t3V\177\000\000\257\034\263\062V\177\000\000\000\000\000\000\000\000\000\000\350rt3V\177\000\000\005", '\000' <repeats 15 times>, "\001\000\000\000\063\061\061\065\300\331t3V\177\000\000\000\000ILV\177\000\000\250a\260mV100\002\000\000\000\060\000\000\000\360\030o1V\177\000\000\000\000\000\000\000\000\000\000\030\335t3V\177\000\000P\002\321-V\177\000\000@\002\321-V\177\000\000\307\016\340=\000\000\000\000\254\241@1V\177\000\000#3\270\b\000\000\000\000\b\370\rEV\177\000\000\210\260\213\062V\177\000\000\070\253\206\063V\177", '\000' <repeats 18 times>, "(\005\321-V\177\000\000\000"...
#2  handle_fatal_signal (signum=6) at global/signal_handler.cc:133
        buf = "*** Caught signal (Aborted) **\n in thread 7f562dd12700 thread_name:ms_async_worker\n\000!\000\000\000\310\034\000X\000\001", '\000' <repeats 18 times>, "\207\341\277\000\000\000\000\000\b\234\020LV\177\000\000\b\000\000\000\000\000\000\000\212s\f\001\000\000\000\000X\223m1V\177\000\000\000\000\000\000\000\000\000\000\370\222m1V\177\000\000\220\231m1V\177\000\000\330\330\rYV\177\000\000\330\330\rYV\177\000\000\331\330\rYV\177\000\000"...
        pthread_name = "ms_async_worker"
#3  <signal handler called>
No symbol table info available.
#4  0x00007f5630b461d7 in raise () from /lib64/libc.so.6
No symbol table info available.
#5  0x00007f5630b478c8 in abort () from /lib64/libc.so.6
No symbol table info available.
#6  0x00007f563144b9d5 in __gnu_cxx::__verbose_terminate_handler() () from /lib64/libstdc++.so.6
No symbol table info available.
#7  0x00007f5631449946 in ?? () from /lib64/libstdc++.so.6
No symbol table info available.
#8  0x00007f5631449973 in std::terminate() () from /lib64/libstdc++.so.6
No symbol table info available.
#9  0x00007f5631449b93 in __cxa_throw () from /lib64/libstdc++.so.6
No symbol table info available.
#10 0x00007f5633375940 in ceph::buffer::list::iterator_impl<false>::advance (this=<optimized out>, o=<optimized out>) at common/buffer.cc:1051
No locals.
#11 0x00007f56334d5359 in AsyncConnection::process (this=0x7f563e56a000) at msg/async/AsyncConnection.cc:814
        bp = {_raw = 0x7f564a18a5e0, _off = 0, _len = 3553509376}
        read = 3553509376
        __func__ = "process"
        __PRETTY_FUNCTION__ = "void AsyncConnection::process()"
        r = <optimized out>
        prev_state = 11
        need_dispatch_writer = false
        l = {mutex = @0x7f563e56a2e0}
#12 0x00007f563347a46f in EventCenter::process_events (this=this@entry=0x7f563e3f82c8, timeout_microseconds=<optimized out>, timeout_microseconds@entry=30000000)
---Type <return> to continue, or q <return> to quit---
    at msg/async/Event.cc:378
        rfired = 1
        event = 0x7f563e4362d0
        cb = 0x7f563e3b6590
        j = <optimized out>
        __PRETTY_FUNCTION__ = "int EventCenter::process_events(int)"
        tv = {tv_sec = 30, tv_usec = 0}
        numevents = 1
        trigger_time = false
        __func__ = "process_events"
        fired_events = std::vector of length 1, capacity 1 = {{fd = 30, mask = 1}}
        now = <optimized out>
#13 0x00007f56334500e0 in Worker::entry (this=0x7f563e3f8280) at msg/async/AsyncMessenger.cc:295
        r = <optimized out>
        __func__ = "entry"
#14 0x00007f563213cdc5 in start_thread () from /lib64/libpthread.so.0
No symbol table info available.
#15 0x00007f5630c0874d in clone () from /lib64/libc.so.6
No symbol table info available.
(gdb) frame 11
#11 0x00007f56334d5359 in AsyncConnection::process (this=0x7f563e56a000) at msg/async/AsyncConnection.cc:814
814	            data_blp.advance(read);
(gdb) p read
$2 = 3553509376
(gdb) p current_header
$3 = {seq = {v = 12133}, tid = {v = 119122}, type = {v = 43}, priority = {v = 127}, version = {v = 7}, front_len = {v = 224}, middle_len = {v = 0}, data_len = {
    v = 3553512290}, data_off = {v = 0}, src = {type = 4 '\004', num = {v = 18}}, compat_version = {v = 2}, reserved = {v = 0}, crc = {v = 3572470833}}
```

david-z commented:
We are not absolutely sure this issue is caused by the MDS reading a large dirfrag. This is just based on our analysis because we don't see any other OPs from MDS could bring back such large data. 

Here is more background:
I check again with our cluster operation engineer today. This issue was only happened on one cluster which was very old and after upgraded from 0.94.x to Jewel. This issue happened 5 times before we fixed it like this PR. 

Only one time we kicked off one client who had a very large directory, then restarting MDS to bring it back. For the other 4 times, we just switched MDS few times, then MDS became available.

batrick commented:
> We are not absolutely sure this issue is caused by the MDS reading a large dirfrag. This is just based on our analysis because we don't see any other OPs from MDS could bring back such large data.

Do you have the MDS log? You could check which osd op was sent to that particular OSD. Then you could inspect the object in question to see how large the omap is.

batrick commented:
But at this point it seems the likely cause. We will need to create a test which checks this.

batrick commented:
@david-z did you modify osd_max_omap_entries_per_request on your cluster? AFAICT, the omap reads should be truncated by default.

gregsfortytwo commented:
Hmm, that option is present in Luminous but not Jewel. I realize now that I don't see version numbers anywhere for what actually experienced the crash?

david-z commented:
> > We are not absolutely sure this issue is caused by the MDS reading a large dirfrag. This is just based on our analysis because we don't see any other OPs from MDS could bring back such large data.
> 
> Do you have the MDS log? You could check which osd op was sent to that particular OSD. Then you could inspect the object in question to see how large the omap is.

Sorry, since MDS log is too much and this issue didn't reproduce under certain pattern, MDS log level on our online clusters is 1/0. We didn't catch the exact OP.

tchaikov commented:
so we are fixing an issue only exists in jewel, and no longer exists in luminous and up. put in other way, it's in an EOL release.

i think the best thing we can do here just have a `ceph_assert(read_len < std::numeric_limits<int>::max())` in `ProtocolV1::handle_message_data()` before passing `read_len` to `bufferlist::iterator::advance()`.

batrick commented:
Agreed with @tchaikov

david-z commented:
Ok, will change this PR accordingly.

david-z commented:
@tchaikov @batrick Pls review it again, thanks.
